client/lib/protect-form/index.jsx

```diff
@@ -57,6 +67,41 @@ export const protectForm = WrappedComponent => {
 	};
 };
 
+/*
+ * Declarative variant that takes a 'isChanged' prop.
+ */
+export class ProtectForm extends Component {
+	static propTypes = {
+		isChanged: PropTypes.bool.isRequired
+	}
+
+	// This function binding makes the listener attached to window unique for each
+	// component instance.
+	warnIfChanged = warnIfChanged.bind();
```

Comment:
I rewrote the code that adds and removes the `beforeunload` listener to add the listener only once, and reference-count multiple instances. It should be much cleaner now, and there is only one `warnIfChanged` function.